src/pub.rs

```diff
@@ -32,11 +32,11 @@ pub(crate) struct PubSocketBackend {
 
 impl PubSocketBackend {
     fn message_received(&self, peer_id: &PeerIdentity, message: Message) {
-        let message = match message {
+        let mut message = match message {
             Message::Message(m) => m,
             _ => return,
         };
-        let data: Vec<u8> = message.into();
+        let data: Vec<u8> = message.pop_front().unwrap().to_vec();
```

Comment:
One of the drawbacks of this approach is that now user can theoretically send and recv message that contain 0 frames. IMHO this is not good and would cause issues with such unwraps. Before user could send an empty message i.e. 1 frame with message of length 0. But currently nothing is preventing him from constructing and empty ZmqMessage and sending it. IMHO we need to think about how this should be handled cause some sockets depend on certain number of frames available (req/rep/router/dealer)

Reply:
I think that should probably be dealt with on the sending side as a message with 0 frames isn't valid per the spec. Currently if we tried to send an empty message I don't think anything would go out on the wire and so we would never get a message that was empty.  It could probably mess up the socket state machine in some cases though so we might want to have some kind of check on the send method.

I think there are two things we could do if you passed an empty message into the send function.
1. Do nothing. Essentially make the send a no-op.  This is what pyzmq does and makes some logical sense in that we asked the library to send 0 data to the other side.
2. Return an Err on the send.  I kind of lean towards this option since if you wanted the empty send to be a no-op you can just handle the Err and do nothing. If you accidentally tried to send an empty message you could catch it.

Reply:
>message with 0 frames isn't valid per the spec.

That's what actually my idea is about.. if 0 frames isn't valid you shouldn't be able to create such a message...
Lets leverage type system to prevent it.

```
    pub fn new() -> Self {
	Self { frames: VecDeque::new() }
    }
```
We should simply not allow empty constuctor to the user. So the message could only be created from a certain data that would be a first frame

Reply:
OK I removed the constructors for empty messages.  This doesn't prevent someone from popping all of the frames from the message and getting an empty message.  Not sure if it is worth trying to prevent that or what the correct thing to do in that case is so it doesn't completely solve the issue of send getting an empty message, but it should hopefully make it harder to do by accident.

Reply:
> This doesn't prevent someone from popping all of the frames from the message and getting an empty message.

Good point.. so there is still a way to misuse the API and we should either prevent it or add some more checks to handle this struct.
@TheButlah what do you think about this MR overall?